[doc] training checkpoint guide with mount_cached (#5100)
skypilot-org/skypilot: docs/source/examples/performance-training.rst

.. _high-performance-training:

High Performance Training
=========================

> This guide covers best practices and examples for achieving high performance distributed training using SkyPilot.

Reviewer: I think it's actually better as is.

Distributed training basics
----------------------------

SkyPilot supports various distributed training frameworks and approaches:

- `PyTorch Distributed Data Parallel (DDP) <https://github.com/skypilot-org/skypilot/tree/master/examples/spot/resnet_ddp>`_
- `Horovod <https://github.com/skypilot-org/skypilot/tree/master/examples/horovod>`_
- `DeepSpeed <https://github.com/skypilot-org/skypilot/tree/master/examples/deepspeed>`_
- `Ray Train <https://github.com/skypilot-org/skypilot/tree/master/examples/ray_train>`_
- `TensorFlow Distribution Strategies <https://github.com/skypilot-org/skypilot/tree/master/examples/tensorflow>`_

The choice of framework depends on your specific needs, but all can be easily configured through SkyPilot's YAML specification.

Best practices
--------------

High performance instances
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Choose high performance instances for optimal training performance. SkyPilot allows you to specify instance types with powerful GPUs and high-bandwidth networking:

- Use the latest GPU accelerators (A100, H100, etc.) for faster training

> - Select instances with NVLink for improved multi-GPU communication

Reviewer 2: As a reader I now wonder "does the spec below select NVLink for me"? Can we look into this & recommend?

> - Consider instances with higher memory bandwidth for large models

Reviewer 2: ```suggestion
- Consider instances with higher memory bandwidth and higher device memory for large models
```

Example configuration:

.. code-block:: yaml

  resources:
    accelerators: 
      A100:1  
      A100-80GB:1  
      H100:1  

Use best disk_tier
~~~~~~~~~~~~~~~~~~~~~~~

> Fast storage is critical for data-intensive workloads in loading and storing data and model checkpoints.

Reviewer 2: ```suggestion
Fast storage is critical for loading and storing data and model checkpoints.
```

> SkyPilot's disk tier options to select the fastest available storage with high-performance local SSDs to reduce I/O bottlenecks

Reviewer 2: ```suggestion
SkyPilot's ``disk_tier`` option supports the fastest available storage with high-performance local SSDs to reduce I/O bottlenecks.
```

Example configuration:

.. code-block:: yaml

  resources:

>     disk_tier: best # using highest performance disk tier

Reviewer 2: ```suggestion
    disk_tier: best  # Use highest performance disk tier.
```

> MOUNT_CACHED for checkpointing

Reviewer 2: ```suggestion
Use ``MOUNT_CACHED`` for checkpointing
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> MOUNT_CACHED provides high performance writing, making it ideal for model checkpoints, logs, and other outputs with fast local writes. Unlike MOUNT mode, it supports all write operations without limitations on file operations. It offers fast file access through a local VFS (Virtual File System) cache implemented by `rclone <https://rclone.org/>`__ that provides near-local disk performance. 

Reviewer 2: ```suggestion
Cloud buckets with the ``MOUNT_CACHED`` mode provides high performance writing, making it ideal for model checkpoints, logs, and other outputs with fast local writes. 

Unlike ``MOUNT`` mode, it supports all write and append operations by using a fast local VFS (Virtual File System) cache implemented by `rclone <https://rclone.org/>`__. Files read from and written to a cloud bucket will go through this cache, which resides on local disk. 
```

Example configuration:

.. code-block:: yaml

    file_mounts:
      /checkpoint:
        name: my-checkpoint-bucket  
        mode: MOUNT_CACHED

.. note::
   

>    In MOUNT_CACHED mode, files only begin uploading after they are closed by all processes. By default, SkyPilot uses sequential transfers to maintain file order, which can cause the cache to grow if files are written faster than they can be uploaded.

Reviewer 2: Discussed:
- move the whole note to Buckets page
- readers may not have a mental model at this point yet to understand 'sequential transfers'

   When multiple jobs share a node (e.g., using fractional GPUs), be aware of these important considerations:
   
   - A job is only marked complete after all files in the cache are flushed to the remote bucket, including those from other jobs.
   - The total disk space required equals the sum of all jobs' cached data.
   - Jobs that finish quickly might need to wait for other jobs' files to be uploaded.

>    - Ensure the disk tier is fast enough to handle the combined write load of all jobs.

Reviewer 2: @SeungjinYang @KeplerC after our discussion, can we simplify this? E.g., one idea is to talk about the behavior for a single job.

> To see the difference between MOUNT and MOUNT_CACHED, see :ref:`storage mounting modes <storage-mounting-modes>`.

Reviewer 2: ```suggestion
For more on the differences between ``MOUNT`` and ``MOUNT_CACHED``, see :ref:`storage mounting modes <storage-mounting-modes>`.
```

Robust checkpointing for spot instances
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When using spot instances, robust checkpointing is crucial for recovering from preemptions. Always try loading from the latest checkpoint first, but have a fallback to earlier checkpoints if the latest one is corrupted.
> Here's a simplified example showing the core concepts:

Reviewer 2: ```suggestion
When using spot instances, robust checkpointing is crucial for recovering from preemptions. 

To make checkpointing reading robust against preemptions, use this recipe:

- Always try loading from the latest checkpoint first 
- If the latest checkpoint is found to be corrupted or incomplete,  fallback to earlier checkpoints

Here's a simplified example showing the core concepts:
```

> .. code-block:: python

Reviewer 2: Discussed:
- We need to subsection this thing, maybe "Adapting ``torch.save``"
- We may need to pull up save_checkpoint
- the example usage is very good, but it's currently buried in docstrs, we may want to pull them out into standalone code snippets.

    disk_tier: best # using highest performance disk tier

  file_mounts:
    /checkpoint:
      name: # NOTE: Fill in your bucket name
      mode: MOUNT_CACHED

  envs:
    # Fill in your wandb key: copy from https://wandb.ai/authorize
    # Alternatively, you can use `--env WANDB_API_KEY=$WANDB_API_KEY`
    # to pass the key in the command line, during `sky jobs launch`.
    WANDB_API_KEY:

  # Assume your working directory is under `~/transformers`.
  workdir: ~/transformers

  setup: |
    pip install -e .
    cd examples/pytorch/question-answering/
    pip install -r requirements.txt torch==1.12.1+cu113 --extra-index-url https://download.pytorch.org/whl/cu113
    pip install wandb

  run: |
    cd examples/pytorch/question-answering/
    python run_qa.py \
      --model_name_or_path bert-base-uncased \
      --dataset_name squad \
      --do_train \
      --do_eval \
      --per_device_train_batch_size 12 \
      --learning_rate 3e-5 \
      --num_train_epochs 50 \
      --max_seq_length 384 \
      --doc_stride 128 \
      --report_to wandb \
      --output_dir /checkpoint/bert_qa/ \
      --run_name $SKYPILOT_TASK_ID \
      --save_total_limit 10 \
      --save_steps 1000

> The highlighted lines add a bucket for checkpoints.

Reviewer 2: Put back highlighted lines ?

Example configuration:

.. code-block:: yaml

    file_mounts:
      /checkpoint:
        name: my-checkpoint-bucket  
        mode: MOUNT_CACHED

.. note::
   
   In MOUNT_CACHED mode, files only begin uploading after they are closed by all processes. By default, SkyPilot uses sequential transfers to maintain file order, which can cause the cache to grow if files are written faster than they can be uploaded.

   When multiple jobs share a node (e.g., using fractional GPUs), be aware of these important considerations:
   
   - A job is only marked complete after all files in the cache are flushed to the remote bucket, including those from other jobs.
   - The total disk space required equals the sum of all jobs' cached data.
   - Jobs that finish quickly might need to wait for other jobs' files to be uploaded.
   - Ensure the disk tier is fast enough to handle the combined write load of all jobs.

To see the difference between MOUNT and MOUNT_CACHED, see :ref:`storage mounting modes <storage-mounting-modes>`.

> Robust checkpointing for spot instances

Reviewer: One other thing that may be helpful - how to I check that my checkpointing/loading code is working? E.g. sky launch, cancel, and launch again.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When using spot instances, robust checkpointing is crucial for recovering from preemptions. 

Reviewer: There is an assumption here: "you should design your job to try and load checkpoints when it starts, because there's no way to know if it's the first run or a restart"
It might be good to clarify the high level thing at the beginning. Specifically:
1. Write checkpoints periodically during iteration code.
2. When you first start, try to load checkpoints. On the first run, this will do nothing, but on restarts it will pick things up where they left off.

My sense is that 2. is slightly unintuitive. If you disagree, maybe it's fine.

Here's a simplified example showing the core concepts:

.. code-block:: python

    def load_checkpoint(save_dir: str):
        try:
            # Find all checkpoints, sorted by step (newest first)
            checkpoints = sorted(
                [f for f in Path(save_dir).glob("checkpoint_*.pt")],
                key=lambda x: int(x.stem.split('_')[-1]),
                reverse=True
            )
            
            # Try each checkpoint from newest to oldest
            for checkpoint in checkpoints:
                try:
                    step = int(checkpoint.stem.split('_')[-1])
                    result = load_checkpoint(checkpoint) # need to fill in
                    return result
                except Exception as e:
                    logger.warning(f"Failed to load checkpoint {step}: {e}")
                    continue
        except Exception as e:
            logger.error(f"Failed to find checkpoints: {e}")
            return None

For a complete implementation with additional features like custom prefixes, extended metadata, and more detailed error handling, see the code below:

.. dropdown:: Full Implementation
    :animate: fade-in-slide-down

>     .. code-block:: python

Reviewer: TBH, I think this is a lot more complicated than it needs to be.
1. The only thing that the `@save_checkpoint` does is write the metadata files, but we don't read them. Is it really necessary?
2. `@load_checkpoint` decorates a function that takes a `step`, but you don't actually need to pass in the `step` when creating it. I'm not sure it makes a lot of sense as a decorator.

To be honest, I think it may be better just to stick to the short snippet above. I'm worried this longer snippet may intimidate people and doesn't provide a lot of value.

Reviewer: This is resolved by https://github.com/skypilot-org/skypilot/pull/5100#discussion_r2025555199

      --save_total_limit 10 \
      --save_steps 1000

The highlighted lines add a bucket for checkpoints.
As HuggingFace has built-in support for periodic checkpointing, we just need to pass the highlighted arguments to save checkpoints to the bucket.
(See more on `Huggingface API <https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments.save_steps>`__).
To see another example of periodic checkpointing with PyTorch, check out `our ResNet example <https://github.com/skypilot-org/skypilot/tree/master/examples/spot/resnet_ddp>`__.

We also set :code:`--run_name` to :code:`$SKYPILOT_TASK_ID` so that the logs for all recoveries of the same job will be saved
to the same run in Weights & Biases.

.. note::
  The environment variable :code:`$SKYPILOT_TASK_ID` (example: "sky-managed-2022-10-06-05-17-09-750781_bert-qa_8-0") can be used to identify the same job, i.e., it is kept identical across all
  recoveries of the job.
  It can be accessed in the task's :code:`run` commands or directly in the program itself (e.g., access
  via :code:`os.environ` and pass to Weights & Biases for tracking purposes in your training script). It is made available to

>   the task whenever it is invoked. See more about :ref:`environment variables provided by SkyPilot <sky-env-vars>`.

Reviewer: Maybe this should be mentioned outside of just the BERT example, like farther up on this page or on the checkpointing section in the managed jobs page.

- `Horovod <https://github.com/skypilot-org/skypilot/tree/master/examples/horovod>`_
- `DeepSpeed <https://github.com/skypilot-org/skypilot/tree/master/examples/deepspeed>`_
- `Ray Train <https://github.com/skypilot-org/skypilot/tree/master/examples/ray_train>`_
- `TensorFlow Distribution Strategies <https://github.com/skypilot-org/skypilot/tree/master/examples/tensorflow>`_

The choice of framework depends on your specific needs, but all can be easily configured through SkyPilot's YAML specification.

Best practices
--------------

High performance instances
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Choose high performance instances for optimal training performance. SkyPilot allows you to specify instance types with powerful GPUs and high-bandwidth networking:

- Use the latest GPU accelerators (A100, H100, etc.) for faster training
- Select instances with NVLink for improved multi-GPU communication
- Consider instances with higher memory bandwidth and higher device memory for large models

Example configuration:

.. code-block:: yaml

  resources:
    accelerators: 
      A100:1  
      A100-80GB:1  
      H100:1  

> Use best disk_tier

Reviewer: ```suggestion
Use ``disk_tier: best``
```
